src/Stratis.Bitcoin.Features.Wallet/Services/WalletService.cs

```diff
@@ -1196,6 +1196,192 @@ public async Task<List<string>> Sweep(SweepRequest request, CancellationToken ca
             }, cancellationToken);
         }
 
+        public async Task<BuildOfflineSignResponse> BuildOfflineSignRequest(BuildOfflineSignRequest request, CancellationToken cancellationToken)
+        {
+            return await Task.Run(() =>
+            {
+                // TODO: It might make more sense to pull out the common code between this and an online transaction build into a common method
+                if (request.Recipients == null)
+                {
+                    request.Recipients = new List<RecipientModel>();
+                }
+
+                if (request.Recipients.Count == 0 && (request.OpReturnAmount == null || request.OpReturnAmount == Money.Zero))
+                    throw new FeatureException(HttpStatusCode.BadRequest, "No recipients.", "Either one or both of recipients and opReturnAmount must be specified.");
+
+                var recipients = new List<Recipient>();
+
+                foreach (RecipientModel recipientModel in request.Recipients)
+                {
+                    if (string.IsNullOrWhiteSpace(recipientModel.DestinationAddress) && string.IsNullOrWhiteSpace(recipientModel.DestinationScript))
+                        throw new FeatureException(HttpStatusCode.BadRequest, "No recipient address.", "Either a destination address or script must be specified.");
+
+                    Script destination;
+
+                    if (!string.IsNullOrWhiteSpace(recipientModel.DestinationAddress))
+                    {
+                        destination = BitcoinAddress.Create(recipientModel.DestinationAddress, this.network).ScriptPubKey;
+                    }
+                    else
+                    {
+                        destination = Script.FromHex(recipientModel.DestinationScript);
+                    }
+
+                    recipients.Add(new Recipient
+                    {
+                        ScriptPubKey = destination,
+                        Amount = recipientModel.Amount
+                    });
+                }
+
+                // If specified, get the change address, which must already exist in the wallet.
+                HdAddress changeAddress = null;
+                if (!string.IsNullOrWhiteSpace(request.ChangeAddress))
```

Comment:
If we have specified a change address, shouldn't this also throw? Or can a null change address be specified?

Reply:
Yes it can be null, in that case the transaction build context will find one from the wallet.